usr/share/rear/rescue/default/500_ssh.sh

```diff
@@ -54,5 +57,110 @@ if has_binary sshd; then
+                cp -p "$original_file" "$target_file"
+
+                # This ssh-keygen invocation checks if a passphrase change succeeds with an empty passphrase for
+                # authentication. If so, the private key file is not currently passphrase-protected.
+                if ssh-keygen -q -p -P '' -N '' -f "$target_file" >/dev/null 2>&1; then
+                    # Handle the unprotected private key file
+
+                    # Obtain a decision on copying and passphrase-protection once
+                    if [ "$passphrase_set" == "no" ]; then
+                        passphrase_set="yes"
+
+                        LogUserOutput ""
+                        LogUserOutput "WARNING: User 'root' on '$HOSTNAME' has potential SSH access to other systems via"
+                        LogUserOutput "unprotected private keys. To copy those keys onto the rescue medium in protected form,"
+                        LogUserOutput "please enter a strong passphrase which will be used when running 'rear recover'."
+                        LogUserOutput "- Entering an empty passphrase will *not* copy unprotected private keys to the rescue medium."
+                        LogUserOutput "- Entering 'UnProtected' (without quotes) will copy unprotected private keys onto the rescue medium."
+                        LogUserOutput ""
+
+                        while true; do
+                            LogUserOutput "SSH client passphrase (or empty, or 'UnProtected'): \c"
+                            read -s passphrase <&6
+                            UserOutput ""
+
+                            case "$passphrase" in
+                                "")
+                                    LogUserOutput "SSH private keys will not be copied"
+                                    break
+                                    ;;
+
+                                UnProtected)
+                                    LogUserOutput ""
+                                    LogUserOutput "WARNING: This will protentially allow anyone in possession of the rescue medium"
+                                    LogUserOutput "unauthenticated access to other systems."
+                                    LogUserOutput ""
+                                    user_input="$( UserInput -I SSH_PRIVATE_KEYS_UNPROTECTED_CONFIRMATION \
+                                        -t 0 -p "Type 'yes' if you want to allow unauthenticated access" )"
+                                    if [ "$user_input" == "yes" ]; then
+                                        LogUserOutput "SSH private keys will be copied in unprotected form"
+                                        break
+                                    fi
+                                    ;;
+
+                                *)
+                                    LogUserOutput "Repeat SSH client passphrase: \c"
+                                    read -s repeated_passphrase <&6
```

Comment:
As above, I wouldn't ask for a password repeat. ReaR is meant to be run in an automated fashion and not manually. I would even go so far as to wonder why we should interactively ask for this password here. Better we should support a variable for the password. If a user really wants to enter this password for *every* `rear mkrescue` run then the user should set the variable via `read` or `UserInput` in the `local.conf`.

Reply:
The primary objective is not to leak any secret into the rescue system. If the passphrase is supplied via a configuration variable in `local.conf` or `site.conf`, it would end up in the rescue system:
```
# ll etc/rear
total 16K
-rw-r--r-- 1 root root 476 Sep 18 23:37 local.conf
-rw-r--r-- 1 root root 275 Sep 18 23:37 os.conf
-rw-r--r-- 1 root root 888 Sep 18 23:37 rescue.conf
-rw-r--r-- 1 root root 137 Sep 18 23:37 site.conf
```

Reply:
Sorry, @jsmeix worked hard to remove all the blocking reads in ReaR. If `UserInput` doesn't yet know how to handle passwords then we should implement this now. IMHO it would be even better to get away without the need to prompt for passwords.

Setting the variable in the configuration doesn't mean that is has to be hard coded. It can be read there with Bash or any other method. If the user puts this configuration into `mkrescue.conf` then it will only be active for the`mkrescue` flow.

Separating between using the passphrase (in these scripts) and providing it (in the configuration) also allows for much greater flexibility. Maybe some user has a secure storage for the passphrase that is not part of the backup. Or maybe he triggers the `rear mkrescue` call from somewhere else and provides the password on the command line or as an environment variable. Lots lots of options open up if we separate those concerns.

Reply:
> Setting the variable in the configuration doesn't mean that is has to be hard coded. It can be read there with Bash or any other method.

This is exactly what I was thinking when I woke up this morning. ;-)

> If the user puts this configuration into 'mkrescue.conf' then it will only be active for the `mkrescue` flow.

I couldn't find any documentation on this. And I'd still prefer to separate secrets from configuration.

> Lots lots of options open up if we separate those concerns.

From my point of view, lots of options are a problem in security, as they make it easy to mess things up with homegrown, poorly understood and poorly tested setups. I'd rather prefer a single, well-understood option which is just "good enough" and works safely even for non-experts.

Reply:
I know that for many ReaR users it takes a bit to get used to the idea that ReaR **configuration** is actually regular **Bash code** that allows them to do what they want.

ReaR has probably thousands of undocumented features :sob: I just happen to remember a lot of the stuff I did 10 years ago.
```bash
~/src/rear/rear$ git grep WORKFLOW.conf
usr/sbin/rear:test -r "$CONFIG_DIR/$WORKFLOW.conf" && Source "$CONFIG_DIR/$WORKFLOW.conf" || true
usr/share/rear/lib/config-functions.sh:    # This is needed, so that they can be overridden in $WORKFLOW.conf
```

If you want to keep using `local.conf` then you can check the `$WORKFLOW` variable yourself:
```bash
if [[ "$WORKFLOW" == mk* ]] ; then
  read -r -s -p "Enter passphrase: " 
  SSH_UNPROTECTED_KEYS_SET_PASSPHRASE="$REPLY" # support passwords with $IFS in them
  read -r -s -p "Confirm passphrase: "
  [ "$REPLY" = "$SSH_UNPROTECTED_KEYS_SET_PASSPHRASE" ] || Error "Passphrase confirmation doesn't match"
fi
```

What do you think about simplifying stuff by removing the SSH key modifications and making do with the improvement of not copying unprotected SSH keys unless specifically configured to do so?

Reply:
@schlomo
only for the fun of it because I just happen to remember
a lot of the stuff I did 10 days or weeks ago ;-)

Plain
<pre>
read -r -s -p "Enter passphrase: "
</pre>
never did work because STDERR is redirected into the log
and 'read' shows its prompt on STDERR.

Nowadays also STDOUT is redirected into the log
so that one should use
<pre>
read -r -s -p "Enter passphrase: " 0<&6 1>&7 2>&8
</pre>
cf. "What to do with stdout and stderr" at
https://github.com/rear/rear/wiki/Coding-Style
The '0<&6' is currently not required but I prefer to use
the whole '0<&6 1>&7 2>&8' term to make it clear that
"all I/O via the original FDs" is meant.

Reply:
Yes, you are right @jsmeix. Can we get a convenience function for that, something like this?
```bash
function without_redirects {
  eval "$@" 0<&6 1>&7 2>&8
}

without_redirects read -r -s -p "Prompt me: "
```

Reply:
Yes we can!
But not today ;-)